### ISSUE TITLE
Lists completely broken - 502 error

### ISSUE THREAD

**Issue by marcusyoung** [bookwyrm-social/bookwyrm#2433]
**Describe the bug**

Lists don't appear to work at all except for the admin user. If another user clicks on them nothing happens. Sends a CPU core up to 100% and eventually goes to a 502 after about a minute. This is for public lists that I created (as admin) if that may be relevant. 

In further test I tried to create a list as a normal (editor) user. It shows the dialogue to give the list a name etc, but then after clicking to create the list it does the same thing. Freezes. Puts a CPU core at 100% and 502s.

I've put all my lists to private until it can be investigated. I have no other user public lists currently.

**Comment by marcusyoung** [comment]
Comment from @jaschaurbach on Matrix
>I have this issue with new lists since 0.5.1
I checked - I had no list created after the update except one. And that fails

**Comment by marcusyoung** [comment]
> Comment from @jaschaurbach on Matrix
> 
> > I have this issue with new lists since 0.5.1
> > I checked - I had no list created after the update except one. And that fails

I think I created most of mine (not the tests) just before 0.5.1, but I have changed the permissions on all of the since 0.5.1. So that may have triggered whatever the issue is.

**Comment by cincodenada** [comment]
Okay yeah I was able to reproduce this too, just accessing a list that I created long before the latest release takes a while and then hits a 502.

Have we tracked down a minimum version here yet? Are we sure this is a regression in the 0.5.1 release, or could it be in 0.5.0 or earlier?

I did a quick scan of the changes and nothing obvious stood out, so I'm doing some further testing now.

**Comment by marcusyoung** [comment]
@cincodenada Had you by any chance changed the permission on that list recently?

**Comment by jaschaurbach** [comment]
I found this in the log

```
2022-11-23 17:13:31,805 DEBG 'bookwyrm' stderr output:
Internal Server Error: /list/2/s/alle-queerwelten-ausgaben
Traceback (most recent call last):
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/core/handlers/base.py", line 204, in _get_response
    response = response.render()
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/response.py", line 83, in rendered_content
    return template.render(context, self._request)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 519, in render
    return self.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 315, in render
    return nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 214, in render
    nodelist.append(node.render_annotated(context))
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/defaulttags.py", line 519, in render
    return self.nodelist.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 988, in render
    output = self.filter_expression.resolve(context)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/django/template/base.py", line 698, in resolve
    new_obj = func(obj, *arg_vals)
  File "/home/queer/bookwyrm/bookwyrm/templatetags/markdown.py", line 13, in get_markdown
    return to_markdown(content)
  File "/home/queer/bookwyrm/bookwyrm/views/status.py", line 284, in to_markdown
    content = markdown(content)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/core.py", line 387, in markdown
    return md.convert(text)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/core.py", line 261, in convert
    self.lines = prep.run(self.lines)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/preprocessors.py", line 81, in run
    parser.close()
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/htmlparser.py", line 78, in close
    super().close()
  File "/usr/lib64/python3.9/html/parser.py", line 114, in close
    self.goahead(1)
  File "/usr/lib64/python3.9/html/parser.py", line 172, in goahead
    k = self.parse_endtag(i)
  File "/usr/lib64/python3.9/html/parser.py", line 420, in parse_endtag
    self.handle_endtag(elem)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/htmlparser.py", line 148, in handle_endtag
    text = self.get_endtag_text(tag)
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/htmlparser.py", line 118, in get_endtag_text
    start = self.line_offset + self.offset
  File "/home/queer/bookwyrm/venv/lib64/python3.9/site-packages/markdown/htmlparser.py", line 95, in line_offset
    return re.match(r'([^\n]*\n){{{}}}'.format(self.lineno-1), self.rawdata).end()
AttributeError: 'NoneType' object has no attribute 'end'
```

**Comment by cincodenada** [comment]
> @cincodenada Had you by any chance changed the permission on that list recently?

I had not, I haven't updated the lists in any way recently

**Comment by marcusyoung** [comment]
> 



> I had not, I haven't updated the lists in any way recently

That's useful to know then. So may go back further. I'd have thought a user would have reported it though and it would have come to light earlier?

**Comment by cincodenada** [comment]
Okay, so I think I figured out why users haven't been reporting this, and it means that this issue is unlikely to be widespread: the issue is only triggered if your user has fewer than 5 books on their shelves (combined). I confirmed this on my `bookwyrm.social` account - it only had one book on the shelf, and I couldn't load lists, I added four more books and now I can.

The admin vs regular user thing is a red herring, I believe - on my instance, my admin user is also my main account, so it has plenty of books on the shelves, whereas the test account I created as a regular user didn't have any books added, and thus triggered the issue. I suspect this is likely to be the case for many of us, but let me know if you see something different!

So, in short, most users won't see this bug because most users have more than 5 books on their shelves. #2437 fixes the underlying issue for users who have fewer than that.

As for what the actual bug was: when we generate suggestions to add to a list, we start by pulling books from the user's shelves that aren't already on the list. Then, if that doesn't get us 5 books, we just pull a few more or less at random from the database at large. This second query is where the bug was, which made it so pulling those "backup" suggestions accidentally queried for all books in the database.

I imagine some users might have still hit this, just because users who are new and didn't import anything (and thus don't have any books on their shelves) are probably likely to go look through lists to find those books, and then those lists won't load.

**Comment by cincodenada** [comment]
I just surveyed all the other slices with an upper bound to make sure this same bug didn't show up elsewhere, and this appears to be the only one, all the other slices are on QuerySet objects directly.

This _might_ be catchable with a pylint rule, it would require some type/object information, but I think pylint has that (which is why it needs to have the actual project, it's not just static analysis iirc)

**Comment by marcusyoung** [comment]
But the issue occurs on list creation as well. So any user who tried to create a list would hit it.

**Comment by marcusyoung** [comment]
Does this also explain why the issue occurs on creating the list? I guess new users are also likely to explore and click on the links at the top to see what they do, so am surprised it hasn't been identified before. How long has this bug existed?

**Comment by marcusyoung** [comment]
And there have been lots of new users lately, so this could potentially explain a lot of the 502s on the big instances. If the server has no CPU resource all users will get 502s whatever they are doing. So many users may have been impacted indirectly.

**Comment by cincodenada** [comment]
> Does this also explain why the issue occurs on creating the list? I guess new users are also likely to explore and click on the links at the top to see what they do, so am surprised it hasn't been identified before. How long has this bug existed?

Yes, list creation is the same issue, as you note the list is created successfully, but list creation then redirects to the new list, which times out. So, it should still only directly affects users with fewer than 5 books on their shelves.

As far as I can tell, this bug has existed since the suggestion feature was added in January 2021. It may be behind some 502s, but I doubt this will fix most of them, since 502 is a pretty generic error that just means the server didn't respond - either because it crashed or, more likely, took too long and timed out (as in this case), and there are a lot of different kinds of things that can take a long time, especially if a server has recently grown significantly.

**Comment by cincodenada** [comment]
In short, I anticipate the brokenness is mostly limited to users without books on their shelves - my server isn't busy enough to have those issues, but looking through my logs, most of the timeouts occur when loading books.

All that said I do think this is a good bug find that is worth a release to get the fix out, thanks for all the info and hunting it down!

**Comment by marcusyoung** [comment]
Thanks. It would be useful for a list of the log files that the application uses to be able to keep an eye on things.

**Comment by hughrun** [comment]
> In short, I anticipate the brokenness is mostly limited to users without books on their shelves - my server isn't busy enough to have those issues, but looking through my logs, most of the timeouts occur when loading books.
> 
OMG this has *definitely* been reported before by new users, and this explains why it's always been new users and then "magically stops happening" once they eventually get a few books on their shelves.

Nice one!

**Comment by mouse-reeve** [comment]
Is this fully fixed by #2437?

**Comment by cincodenada** [comment]
It should be, yes - #2437 was the actual fix, the CPU usage (and subsequent breakage) is explained by the un-limited query, so limiting it should prevent that :+1:
